File: mobile/src/main/java/rhedox/gesahuvertretungsplan/ui/activity/MainActivity.java
package rhedox.gesahuvertretungsplan.ui.activity;

import android.annotation.TargetApi;
import android.app.ActivityManager;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.ColorInt;
import android.support.annotation.DimenRes;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TabLayout;
import android.support.v4.util.Pair;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.support.v7.widget.Toolbar;
import android.content.Intent;
import android.view.View;
//import android.widget.DatePicker;

import com.afollestad.materialcab.MaterialCab;

import org.joda.time.DateTime;
import org.joda.time.DateTimeConstants;
import org.joda.time.LocalDate;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import rhedox.gesahuvertretungsplan.R;
import rhedox.gesahuvertretungsplan.model.SchoolWeek;
import rhedox.gesahuvertretungsplan.model.Substitute;
import rhedox.gesahuvertretungsplan.ui.adapters.PagerAdapter;
import rhedox.gesahuvertretungsplan.ui.fragment.AnnouncementFragment;
import rhedox.gesahuvertretungsplan.ui.fragment.DatePickerFragment;
import rhedox.gesahuvertretungsplan.ui.fragment.MainFragment;
import rhedox.gesahuvertretungsplan.ui.fragment.PreferenceFragment;
import rhedox.gesahuvertretungsplan.util.SubstituteShareHelper;
import rhedox.gesahuvertretungsplan.util.TabLayoutHelper;

public class MainActivity extends AppCompatActivity implements ViewPager.OnPageChangeListener, MaterialCab.Callback, MainFragment.MaterialActivity {
    private boolean sortImportant;
    private boolean specialMode;
	private boolean isAmoledBlackEnabled;

    private boolean canGoBack = false;

    @BindView(R.id.coordinator) CoordinatorLayout coordinatorLayout;

    private PagerAdapter pagerAdapter;

    @BindView(R.id.fab) FloatingActionButton floatingActionButton;

    @BindView(R.id.appbarLayout) AppBarLayout appBarLayout;
    private AppBarLayout.OnOffsetChangedListener appBarLayoutOffsetListener;
    @BindView(R.id.viewPager) ViewPager viewPager;
    @BindView(R.id.tabLayout) TabLayout tabLayout;

    @BindView(R.id.toolbar) Toolbar toolbar;
    private Unbinder unbinder;

    public static final String EXTRA_DATE ="date";
    public static final String EXTRA_WIDGET ="widget";

    //Store Date of Monday of current week
    private LocalDate date;

    private MainFragment currentFragment;

    private MaterialCab cab;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //Preferences
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);

        sortImportant = prefs.getBoolean(PreferenceFragment.PREF_SORT, false);
        specialMode = prefs.getBoolean(PreferenceFragment.PREF_SPECIAL_MODE, false);

        boolean isWhiteIndicatorEnabled = prefs.getBoolean(PreferenceFragment.PREF_WHITE_TAB_INDICATOR, false);
        isAmoledBlackEnabled = prefs.getBoolean(PreferenceFragment.PREF_AMOLED, false);

        TypedArray typedArray = getTheme().obtainStyledAttributes(new int[]{R.attr.colorAccent});
        int color = typedArray.getColor(0, 0xFFFFFFFF);
        typedArray.recycle();

        //Initialize UI
        setTheme(R.style.GesahuTheme);
        setContentView(R.layout.activity_main);
        unbinder = ButterKnife.bind(this);
        setSupportActionBar(toolbar);

        if(isAmoledBlackEnabled)
            getWindow().setBackgroundDrawable(new ColorDrawable(Color.BLACK));

        //Use AppBarLayout with SwipeRefreshLayout
        appBarLayoutOffsetListener = new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                if(currentFragment != null)
                    currentFragment.setSwipeToRefreshEnabled(verticalOffset == 0);
            }
        };
        appBarLayout.addOnOffsetChangedListener(appBarLayoutOffsetListener);

        //Add back button and tab bar inset
        Bundle extras = getIntent().getExtras();
        if(extras != null && extras.containsKey(EXTRA_DATE)) {
            date = new DateTime(getIntent().getExtras().getLong(EXTRA_DATE)).toLocalDate();

            if(!extras.containsKey(EXTRA_WIDGET) || !extras.getBoolean(EXTRA_WIDGET)) {
                canGoBack = true;
                getSupportActionBar().setDisplayHomeAsUpEnabled(true);

                toolbar.setContentInsetsRelative((int) getResources().getDimension(R.dimen.default_content_inset), toolbar.getContentInsetEnd());
            }
        }
        else
            date = SchoolWeek.next();

        setupViewPager(date, isWhiteIndicatorEnabled ? 0xFFFFFFFF : color, savedInstanceState != null);

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
            setupTaskDescription();

        if (savedInstanceState != null) {
            // Restore the CAB state, save a reference to cab.
            cab = MaterialCab.restoreState(savedInstanceState, this, this);
        } else {
            cab = new MaterialCab(this, R.id.cab_stub);
            cab.setMenu(R.menu.menu_cab_main);
            cab.setTitleRes(R.string.main_cab_title);
        }
    }

    private void setupViewPager(LocalDate date, @ColorInt int indicatorColor, final boolean isRestored)
    {
        final Pair<LocalDate, Integer> pair = MainActivity.getDate(date);

        pagerAdapter = new PagerAdapter(getSupportFragmentManager(), pair.first);

        viewPager.setAdapter(pagerAdapter);
        viewPager.addOnPageChangeListener(this);

        tabLayout.setupWithViewPager(viewPager);
        tabLayout.setSelectedTabIndicatorColor(indicatorColor);
        if(canGoBack)
            TabLayoutHelper.setContentInsetStart(tabLayout, (int)getResources().getDimension(R.dimen.default_content_inset));

        if(!isRestored)
            viewPager.setCurrentItem(pair.second);

        //shitty workaround for calling onPageSelected on first page
        //WHY DO YOU DO THIS ANDROID? WHY?!
        viewPager.post(new Runnable() {
            @Override
            public void run() {
                if (viewPager.getCurrentItem() == 0 || currentFragment == null)
                    onPageSelected(viewPager.getCurrentItem());
            }
        });
    }

    private static Pair<LocalDate, Integer> getDate(LocalDate date)
    {
        int index = date.getDayOfWeek() - DateTimeConstants.MONDAY;
        date = date.minusDays(index);
        int dayIndex = Math.max(0, Math.min(index, 5));

        return new Pair<LocalDate, Integer>(date, dayIndex);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    private void setupTaskDescription()
    {
        TypedArray a = obtainStyledAttributes(new int[]{R.attr.colorPrimary});
        int primaryColor = a.getColor(0, 0);
        a.recycle();

        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.ic_task);
        ActivityManager.TaskDescription description = new ActivityManager.TaskDescription(getString(R.string.app_name), bitmap, primaryColor);
        this.setTaskDescription(description);
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.fab)
    public void showAnnouncement(View view) {
        if(currentFragment != null && currentFragment.getSubstitutesList() != null && currentFragment.getSubstitutesList().hasAnnouncement())
            AnnouncementFragment.newInstance(currentFragment.getSubstitutesList().getAnnouncement()).show(getSupportFragmentManager(), AnnouncementFragment.TAG);
    }

    @Override
    protected void onResume() {
        super.onResume();

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getBaseContext());
        boolean previouslyStarted = prefs.getBoolean(PreferenceFragment.PREF_PREVIOUSLY_STARTED, false);
        if(!previouslyStarted) {
            Intent intent = new Intent(this, WelcomeActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            startActivity(intent);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent TargetActivity in AndroidManifest.xml.

        switch(item.getItemId()) {
            case R.id.action_settings:
                Intent intent = new Intent(this, PreferenceActivity.class);
                startActivity(intent);
                break;

            case R.id.action_load:
                LocalDate pickerDate;
                if(currentFragment != null && currentFragment.getDate() != null)
                    pickerDate = currentFragment.getDate();
                else if(this.date != null)
                    pickerDate = this.date;
                else
                    pickerDate = LocalDate.now();

                DatePickerFragment.newInstance(pickerDate).show(getSupportFragmentManager(), DatePickerFragment.TAG);
                break;

            case R.id.action_about:
                AboutLibs.start(this, isAmoledBlackEnabled);
                break;

            case android.R.id.home:
                if(canGoBack)
                    this.onBackPressed();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        unbinder.unbind();

        pagerAdapter.destroy();
        if(appBarLayout != null)
            appBarLayout.removeOnOffsetChangedListener(appBarLayoutOffsetListener);

        if(viewPager != null)
            viewPager.removeOnPageChangeListener(this);

        if(floatingActionButton != null)
            floatingActionButton.setOnClickListener(null);

        super.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        // If the CAB isn't null, save it's state for restoration in onCreate()
        if (cab != null)
            cab.saveState(outState);
    }

    @Override
    public void onPageSelected(int position) {
        if(pagerAdapter == null || floatingActionButton == null)
            return;

        currentFragment = pagerAdapter.getFragment(getSupportFragmentManager(), position);

        if(currentFragment != null)
            currentFragment.onDisplay();

        updateUI();
    }


    //Called when the displayed substitutes of the visible fragment change or a substitute gets selected
    public void updateUI() {
        updateFabVisibility();
        updateCabVisibility();
        updateTabInset();

        //Expand app bar when there are no substitutes
        if(currentFragment != null && (currentFragment.getSubstitutesList() == null || !currentFragment.getSubstitutesList().hasSubstitutes()))
            expandAppBar();
    }

    //Shows a specific date
    public void showDate(LocalDate date) {
        if(date.getWeekOfWeekyear() != this.date.getWeekOfWeekyear()) {
            //Launch a new activity with that week
            Intent intent = new Intent(this.getApplicationContext(), MainActivity.class);
            intent.putExtra(MainActivity.EXTRA_DATE, date.toDateTimeAtCurrentTime().getMillis());
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(intent);
        } else {
            //Same week => just switch to day-tab
            int index = date.getDayOfWeek() - DateTimeConstants.MONDAY;
            int dayIndex = Math.max(0, Math.min(index, 5));
            if(viewPager != null)
                viewPager.setCurrentItem(dayIndex);
        }
    }


    @Override
    public CoordinatorLayout getCoordinatorLayout() {
        return coordinatorLayout;
    }


    private void updateFabVisibility() {
        if(floatingActionButton != null) {
            boolean isVisible = false;

            //Show fab only if there is an announcement and no substitute is selected
            if(currentFragment != null)
                isVisible = currentFragment.getSubstitutesList() != null && currentFragment.getSubstitutesList().hasAnnouncement() && (currentFragment.getAdapter() == null || currentFragment.getAdapter().getSelectedIndex() == -1);

            if(isVisible) {
                floatingActionButton.setEnabled(true);
                floatingActionButton.show();
            } else {
                floatingActionButton.hide();
                floatingActionButton.setEnabled(false);
            }
        }
    }
    private void updateCabVisibility() {
        if(cab != null) {

            //Show cab if a substitute is selected and it's not visible yet (pls no stackoverflow)
            if(currentFragment != null && currentFragment.getAdapter() != null && currentFragment.getAdapter().getSelectedIndex() != -1) {
                if(!cab.isActive()) {
                    cab.start(this);
                    expandAppBar();
                }

            } else {
                if(cab.isActive())
                    cab.finish();
            }
        }
    }
    private void updateTabInset() {
        //Inset the tabs if we can go back or a substitute is selected (back button is visible)
        @DimenRes int res = cab != null && cab.isActive() || canGoBack ? R.dimen.default_content_inset : R.dimen.default_content_inset_small;
        TabLayoutHelper.setContentInsetStart(tabLayout, (int) getResources().getDimension(res));

    }
    private void expandAppBar() {
        if(appBarLayout != null)
            appBarLayout.setExpanded(true, true);
    }

    public MainFragment getVisibleFragment() {
        return currentFragment;
    }

    @Override
    public boolean onCabItemClicked(MenuItem item) {
        if(item == null || currentFragment == null || currentFragment.getAdapter() == null)
            return false;

        Substitute substitute = currentFragment.getAdapter().getSelected();
        if(substitute == null)
            return false;

        switch(item.getItemId()) {

            case R.id.action_share:
                    LocalDate date;
                    if(currentFragment.getSubstitutesList() != null)
                        date = currentFragment.getSubstitutesList().getDate();
                    else
                        date = null;

                    startActivity(SubstituteShareHelper.makeShareIntent(this, date, substitute));
                return true;
        }

        return false;
    }

    @Override
    public boolean onCabFinished(MaterialCab cab) {
        if(currentFragment != null && currentFragment.getAdapter() != null)
            currentFragment.getAdapter().clearSelection();

        return true;
    }

//region Unused interface methods
    @Override
    public void onPageScrollStateChanged(int state) {}
    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

    @Override
    public boolean onCabCreated(MaterialCab cab, Menu menu) {
        return true;
    }
    //endregion
}

//https://code.google.com/p/android/issues/detail?id=78062

File: mobile/src/main/java/rhedox/gesahuvertretungsplan/ui/activity/MainActivity.kt
package rhedox.gesahuvertretungsplan.ui.activity

import android.Manifest
import android.accounts.Account
import android.annotation.TargetApi
import android.app.ActivityManager
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.edit
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import com.google.android.material.navigation.NavigationView
import com.google.firebase.analytics.FirebaseAnalytics
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.clearTask
import org.jetbrains.anko.displayMetrics
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.newTask
import org.joda.time.LocalDate
import rhedox.gesahuvertretungsplan.App
import rhedox.gesahuvertretungsplan.R
import rhedox.gesahuvertretungsplan.model.database.entity.Board
import rhedox.gesahuvertretungsplan.mvp.MainContract
import rhedox.gesahuvertretungsplan.presenter.MainPresenter
import rhedox.gesahuvertretungsplan.presenter.state.MainState
import rhedox.gesahuvertretungsplan.service.CalendarSyncService
import rhedox.gesahuvertretungsplan.service.GesaHuAccountService
import rhedox.gesahuvertretungsplan.ui.fragment.*
import rhedox.gesahuvertretungsplan.util.accountManager
import rhedox.gesahuvertretungsplan.util.isCalendarWritingPermissionGranted
import rhedox.gesahuvertretungsplan.util.localDateFromUnix
import javax.inject.Inject

/**
 * Created by robin on 20.10.2016.
 */
class MainActivity : AppCompatActivity(), MainContract.View, DrawerActivity, NavigationActivity {
    private var toggle: ActionBarDrawerToggle? = null
    private lateinit var listener: Listener
    private var drawerSelected: Int? = null
    private lateinit var analytics: FirebaseAnalytics
    private var isAmoledBlackEnabled = false

    private lateinit var headerUsername: TextView

    private lateinit var currentFragment: Fragment

    private lateinit var presenter: MainContract.Presenter

    @Inject internal lateinit var prefs: SharedPreferences

    override var userName: String
        get() = headerUsername.text.toString()
        set(value) {
            headerUsername.text = value
        }

    override var currentDrawerId: Int = -1
        set(value) {
            field = value
            val menuItem: MenuItem? = when (value) {
                MainContract.DrawerIds.settings -> navigationView.menu.findItem(R.id.settings)
                MainContract.DrawerIds.about -> navigationView.menu.findItem(R.id.about)
                MainContract.DrawerIds.substitutes -> navigationView.menu.findItem(R.id.substitutes)
                else -> navigationView.menu.findItem(value)
            }
            if(menuItem != null) {
                menuItem.isChecked = true
            }
        }

    override fun getIsPermanentDrawer(): Boolean {
        return drawer == null
    }

    companion object {
        const val currentFragmentTag = "pageFragment"
        const val state = "navPresenterState"
        const val calendarPermissionRequestCode = 9
    }

    object Extra {
        const val date = "date"
    }

    object Action {
        const val calendarPermission = "calendarPermission"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val appComponent = (application as App).appComponent
        appComponent.inject(this)

        isAmoledBlackEnabled = prefs.getBoolean(PreferenceFragment.PREF_AMOLED, false)

        analytics = FirebaseAnalytics.getInstance(this)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setupTaskDescription()
        }

        //Restore presenter
        val state = savedInstanceState?.getParcelable(MainActivity.state) ?: MainState()
        presenter = if (lastCustomNonConfigurationInstance != null) {
            lastCustomNonConfigurationInstance as MainPresenter
        } else {
            MainPresenter(appComponent.plusBoards(), state)
        }

        //Setup view
        if (isAmoledBlackEnabled)
            this.setTheme(R.style.GesahuThemeAmoled)
        else
            this.setTheme(R.style.GesahuTheme)

        setContentView(R.layout.activity_main)
        setupDrawerLayout()

        headerUsername = navigationView.getHeaderView(0).findViewById(R.id.headerUsername)
        navigationView.setNavigationItemSelectedListener {
            drawerSelected = it.itemId
            if (drawer != null) {
                drawer?.closeDrawer(GravityCompat.START)
            } else {
                onDrawerItemSelected(drawerSelected!!)
            }
            true
        }

        //Show initial fragment
        val fragment = supportFragmentManager.findFragmentByTag(currentFragmentTag)
        val timestamp: Int? = intent?.getIntExtra(Extra.date, 0)
        val date = localDateFromUnix(timestamp)
        if (fragment != null && date == null) {
            currentFragment = fragment
        } else {
            currentFragment = SubstitutesFragment.newInstance(date)
            supportFragmentManager.beginTransaction().replace(R.id.fragment_container, currentFragment, currentFragmentTag).commit()
        }

        if (intent?.action == Action.calendarPermission) {
            askForCalendarPermission()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(state, presenter.saveState())
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    @Suppress("DEPRECATION")
    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    private fun setupTaskDescription() {
        val a = obtainStyledAttributes(intArrayOf(R.attr.colorPrimary))
        val primaryColor = a.getColor(0, 0)
        a.recycle()

        val description = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            ActivityManager.TaskDescription(getString(R.string.app_name), R.drawable.ic_task, primaryColor)
        } else {
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.ic_task)
            ActivityManager.TaskDescription(getString(R.string.app_name), bitmap, primaryColor)
        }
        this.setTaskDescription(description)
    }

    override fun onResume() {
        super.onResume()
        val menuItem = navigationView.menu.findItem(currentDrawerId)
        if(menuItem != null) {
            menuItem.isChecked = true
        }
    }

    private fun setupDrawerLayout() {
        if (drawer != null) {
            toggle = ActionBarDrawerToggle(this, drawer, R.string.drawer_open, R.string.drawer_close)
            toggle!!.drawerArrowDrawable.color = Color.WHITE
            toggle!!.setHomeAsUpIndicator(null)
            toggle!!.isDrawerIndicatorEnabled = true
            drawer!!.addDrawerListener(toggle!!)

            listener = Listener()
            listener.callback = {
                if (drawerSelected != null) {
                    onDrawerItemSelected(drawerSelected!!)
                }
            }
            drawer!!.addDrawerListener(listener)
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.statusBarColor = 0

            if (drawer == null) {
                rootLayout?.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
            }
        }
    }

    private fun onDrawerItemSelected(id: Int) {
        when (id) {
            R.id.about -> presenter.onNavigationDrawerItemClicked(MainContract.DrawerIds.about)
            R.id.settings -> presenter.onNavigationDrawerItemClicked(MainContract.DrawerIds.settings)
            R.id.substitutes -> presenter.onNavigationDrawerItemClicked(MainContract.DrawerIds.substitutes)
            R.id.supervisions -> presenter.onNavigationDrawerItemClicked(MainContract.DrawerIds.supervisions)
            else -> presenter.onNavigationDrawerItemClicked(drawerSelected!!)
        }
    }

    override fun syncDrawer() {
        toggle?.syncState()
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        toggle?.syncState()
    }

    override fun onDestroy() {
        super.onDestroy()

        if (!isChangingConfigurations) {
            presenter.destroy()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        toggle?.onConfigurationChanged(newConfig)
    }

    override fun onRetainCustomNonConfigurationInstance(): Any {
        return presenter
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        if (toggle?.onOptionsItemSelected(item) == true) {
            return true
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onBackPressed() {
        if (drawer?.isDrawerOpen(GravityCompat.START) == true)
            drawer!!.closeDrawer(GravityCompat.START)
        else
            super.onBackPressed()
    }

    override fun showBoards(boards: List<Board>) {
        val menu = navigationView.menu
        menu.removeGroup(R.id.boardsSubheader)
        for (board in boards) {
            val item = menu.add(R.id.boardsSubheader, board.name.hashCode() + MainContract.DrawerIds.board, Menu.NONE, board.name)
            item.isCheckable = true
            item.isChecked = currentDrawerId == board.name.hashCode() + MainContract.DrawerIds.board
        }
    }

    override var avatar: Bitmap? = null
        set(value) {
            field = value

            val imageView = navigationView.getHeaderView(0).findViewById<CircleImageView>(R.id.avatarView)
            if(field != null) {
                imageView.setPadding(0, 0, 0, 0)
                imageView.setImageBitmap(avatar)
            } else {
                val padding = (displayMetrics.density * 4).toInt()
                imageView.setPadding(padding, padding, padding, padding)
                imageView.setImageResource(R.drawable.ic_person)
            }
        }

    override fun navigateToSettings() {
        (currentFragment as? AnimationFragment)?.useSlideAnimation = true
        val fragment = PreferenceContainerFragment.newInstance()
        fragment.useSlideAnimation = true
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, currentFragmentTag)
                .commit()
        currentFragment = fragment
    }

    override fun navigateToAbout() {
        (currentFragment as? AnimationFragment)?.useSlideAnimation = true
        val fragment = AboutContainerFragment.newInstance()
        fragment.useSlideAnimation = true
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, currentFragmentTag)
                .commit()
        this.currentFragment = fragment
        title = getString(R.string.action_about)
    }

    override fun navigateToIntro() {
        startActivity(intentFor<WelcomeActivity>().newTask().clearTask())
    }

    override fun navigateToBoard(boardName: String) {
        (currentFragment as? AnimationFragment)?.useSlideAnimation = true
        val fragment = BoardFragment.newInstance(boardName)
        fragment.useSlideAnimation = true
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, currentFragmentTag)
                .commit()
        this.currentFragment = fragment
    }

    override fun navigateToAuth() {
        accountManager.addAccount(GesaHuAccountService.GesaHuAuthenticator.accountType,
                null, null, null, this, null, null)
    }

    override fun navigateToSubstitutes(date: LocalDate?) {
        (currentFragment as? AnimationFragment)?.useSlideAnimation = currentFragment !is SubstitutesFragment
        val fragment = SubstitutesFragment.newInstance(date)
        fragment.useSlideAnimation = currentFragment !is SubstitutesFragment
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, currentFragmentTag)
                .commit()
        this.currentFragment = fragment
    }

    override fun navigateToSupervisions(date: LocalDate?) {
        (currentFragment as? AnimationFragment)?.useSlideAnimation = currentFragment !is SupervisionsFragment
        val fragment = SupervisionsFragment.newInstance(date)
        fragment.useSlideAnimation = currentFragment !is SupervisionsFragment
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, currentFragmentTag)
                .commit()
        this.currentFragment = fragment
    }

    private fun askForCalendarPermission() {
        if (isCalendarWritingPermissionGranted) return
        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_CALENDAR, Manifest.permission.READ_CALENDAR), calendarPermissionRequestCode)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (requestCode != calendarPermissionRequestCode) return
        presenter.onCalendarPermissionResult(permissions.firstOrNull() == Manifest.permission.WRITE_CALENDAR && grantResults.firstOrNull() == PackageManager.PERMISSION_GRANTED)
    }

    override fun updateCalendarSync(account: Account) {
        prefs.edit {
            this.putBoolean(CalendarSyncService.alreadyAskedForCalendarPreference, true)
        }
        CalendarSyncService.updateIsSyncable(account, applicationContext, prefs)
    }

    class Listener: DrawerLayout.SimpleDrawerListener() {
        var callback: ((drawer: NavigationView) -> Unit)? = null

        override fun onDrawerClosed(drawerView: View) {
            super.onDrawerClosed(drawerView)
            callback?.invoke(drawerView as NavigationView)
        }
    }
}